core/src/main/java/org/apache/iceberg/FetchPlanningResultResponseParser.java

```diff
@@ -0,0 +1,177 @@
+    }
+
+    if ((response.deleteFiles() != null && !response.deleteFiles().isEmpty())
+        && (response.fileScanTasks() == null || response.fileScanTasks().isEmpty())) {
+      throw new IllegalArgumentException(
+          "Invalid response: deleteFiles should only be returned with fileScanTasks that reference them");
+    }
+
+    gen.writeStartObject();
+    if (response.planStatus() != null) {
+      gen.writeStringField(PLAN_STATUS, response.planStatus().status());
+    }
+
+    if (response.planTasks() != null) {
+      gen.writeArrayFieldStart(PLAN_TASKS);
+      for (String planTask : response.planTasks()) {
+        gen.writeString(planTask);
+      }
+      gen.writeEndArray();
+    }
+
+    List<DeleteFile> deleteFiles = null;
+    Map<String, Integer> deleteFilePathToIndex = Maps.newHashMap();
+    if (response.deleteFiles() != null) {
+      deleteFiles = response.deleteFiles();
+      gen.writeArrayFieldStart(DELETE_FILES);
+      for (int i = 0; i < deleteFiles.size(); i++) {
+        DeleteFile deleteFile = deleteFiles.get(i);
+        deleteFilePathToIndex.put(String.valueOf(deleteFile.path()), i);
+        ContentFileParser.unboundContentFileToJson(
+            deleteFiles.get(i), response.specsById().get(deleteFile.specId()), gen);
+      }
+      gen.writeEndArray();
+    }
+
+    if (response.fileScanTasks() != null) {
+      Set<Integer> deleteFileReferences = Sets.newHashSet();
+      gen.writeArrayFieldStart(FILE_SCAN_TASKS);
+      for (FileScanTask fileScanTask : response.fileScanTasks()) {
+        if (deleteFiles != null) {
+          for (DeleteFile taskDelete : fileScanTask.deletes()) {
+            deleteFileReferences.add(deleteFilePathToIndex.get(taskDelete.path().toString()));
+          }
+        }
+        PartitionSpec partitionSpec = response.specsById().get(fileScanTask.file().specId());
+        RESTFileScanTaskParser.toJson(fileScanTask, deleteFileReferences, partitionSpec, gen);
+      }
+      gen.writeEndArray();
+    }
+
+    gen.writeEndObject();
+  }
+
+  public static FetchPlanningResultResponse fromJson(String json) {
+    Preconditions.checkArgument(json != null, "Invalid response: fetchPanningResultResponse null");
+    return JsonUtil.parse(json, FetchPlanningResultResponseParser::fromJson);
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static FetchPlanningResultResponse fromJson(JsonNode json) {
+    Preconditions.checkArgument(
+        json != null && !json.isEmpty(),
+        "Invalid response: fetchPanningResultResponse null or empty");
+
+    PlanStatus planStatus = null;
+    if (json.has(PLAN_STATUS)) {
+      String status = JsonUtil.getString(PLAN_STATUS, json);
+      planStatus = PlanStatus.fromName(status);
+    }
+
+    List<String> planTasks = JsonUtil.getStringListOrNull(PLAN_TASKS, json);
+
+    List<DeleteFile> allDeleteFiles = null;
+    if (json.has(DELETE_FILES)) {
+      JsonNode deletesArray = json.get(DELETE_FILES);
```

Comment:
you need to be careful here and in a few other places. This node could still be null or a non-array, so you need to check for that too.
It's better to use `JsonUtil.get(...)` (since that will also check for potential nulls) + a check that makes sure this is an array: `Preconditions.checkArgument(deletesArray.isArray(), ...)`(similar to https://github.com/apache/iceberg/blob/fb929437af7f25777d33f5dfe6d14698f09e511e/core/src/main/java/org/apache/iceberg/rest/requests/UpdateTableRequestParser.java#L101-L103).

Please make sure to also update the other places in this and in other parsers

Reply:
Cleaning this up, also we should validate the actual content in the content file instead of directly attempting a cast